### PR TITLE
Feature: add initial GA4 support

### PR DESCRIPTION
**Pull request by khawkins98**
This addresses https://github.com/visual-framework/vf-core/issues/1428 and supports classic GA and GA4. I've done some testing, but further testing is recommended (note I've suggested this as 1.1.0-rc.0)

### CODE REVIEW

**Comment by sandykadam** [comment]
Thanks @khawkins98, this is great. In-fact I was going to start discussion with team to prioritise this as G4 rolling out next year.

Will update you on testing of this.

**Comment by nitinja** [comment]
Great, accepted, thanks

File: components/vf-componenet-rollup/scripts.js [hunk 1]
```diff
@@ -20,7 +20,8 @@ vfBannerElixir();
 
 import { vfGaIndicateLoaded } from "vf-analytics-google/vf-analytics-google";
 let vfGaTrackOptions = {
-  vfGaTrackPageLoad: true
+  vfGaTrackPageLoad: true,
+  vfGa4MeasurementId: "G-N5ZLH2KNJP"
```
Comment (sandykadam) [inline components/vf-componenet-rollup/scripts.js]: Is this intentional or left by mistake.

Comment (khawkins98) [reply components/vf-componenet-rollup/scripts.js]: This is used by the visual-framework.dev doc site, but it probably raises more chances for confusion than help. So I made a clear generic that needs to be updated.

File: components/vf-analytics-google/vf-analytics-google.js [hunk 2]
```diff
@@ -116,21 +122,30 @@ function vfGetMeta(metaName) {
...
-
-  // Need help
+  // Need help?
   // How to add dimension to your property
   // https://developers.google.com/analytics/devguides/collection/analyticsjs/custom-dims-mets
   // https://support.google.com/analytics/answer/2709829?hl=en
 
+  if (typeof gtag === "undefined") {
+    // if the site is still using legacy GA, set a dummy gtag function so we don't have to add a bunch of if statements
+    var gtag = function() {};
```
Comment (oss6) [inline components/vf-analytics-google/vf-analytics-google.js]: I encountered an issue when testing this. When using `var` the `gtag` is hoisted, hence it'll be `undefined` before the check (see https://stackoverflow.com/a/36911203).
A solution is to use the following `window.gtag = function() {}`

Comment (khawkins98) [reply components/vf-analytics-google/vf-analytics-google.js]: Hi Ossama! By pure coincidence I've been working through this code this morning, I'll be revising this PR with some fixes that include this issue.

File: components/vf-analytics-google/vf-analytics-google.js [hunk 3]
```diff
@@ -189,6 +209,7 @@ function vfGaInit(vfGaTrackOptions) {
   // standard google analytics bootstrap
   if (vfGaTrackOptions.vfGaTrackPageLoad) {
     ga("send", "pageview");
```
Comment (oss6) [inline components/vf-analytics-google/vf-analytics-google.js:226]: Here `ga` may not be defined if one moves to ga4. I think we may need to use the same checks for `gtag`

**Comment by khawkins98** [comment]
I've just pushed a fair number of changes (broken down into separate thematic commits) -- these address some issues that @oss6 spotted, in addition to making it a bit more resilient, improving the optional debug command and sending more appropriate events to GA4.

There's also a documentation task to be done, as GA4 has higher expectations on configuring the Google Analytics Profile for custom events (see section "IMPORTANT: Register custom definitions" at https://www.analyticsmania.com/post/how-to-track-custom-events-with-google-analytics-4/#custom-definitions) 

~I also want to point out in the readme that this approach does not work well at all when attempting to combine it with GA4 that is loaded through Google Tag Manager -- you should use the local JS approach:~ (I was wrong, see comment below)

**Comment by khawkins98** [comment]
> I also want to point out in the readme that this approach does not work well at all when attempting to combine it with GA4 that is loaded through Google Tag Manager -- you should use the local JS approach:

I misspoke about that, Google Tag Manager should work fine. In my case, there was another bug causing the two clash.

**Comment by khawkins98** [comment]
I'm still working my way through some of the complication between GA4 + UA + gtag + Google Tag Manager and am posting some things here before I forget them over the long weekend here:

1. In _principle_ you can use this gtag.js method with Google Tag Manager, but it seems unreliable and often leads to race conditions where this code may fail to work -- I think it's not realistic to support using this + Google Tag Manager. BUT, this is consistent with most of my reading on best practice. It is summed up well at https://measureschool.com/google-tag-manager-vs-global-site-tag/
    - > However, both these tools utilize some common resources that can lead to conflict in some cases. For example, the data layer is one such resource that is used by both tools. Therefore, it is important to be careful while adding both of them together. You will also need to test the implementation of tools by checking whether the correct data is sent to these tools. Overall, the process is doable but complicated. That’s why I would recommend not to install them together on the same page. 
    - That is: if you want to use Google Tag Manager, you should use that for your events. If you want to use gtag.js + custom JS, then vf-google-analytics is good for you. This code will provide particular value if you need a consistent way to collect event data across many websites, which is what we're after.
2. The basics of `vf-analytics-google` will be plug-and-play if using gtag+GA4, but to get the full value out if you need to add (currently) 4 "custom definitions": 
    -  grafik
    - This is an important step that will need to be done on each GA4 profile (adding the Custom Definitions will not retroactively collect data).
3. Good news: None of the changes in this PR should result in regressions for sites using UA + ga.js

This all needs to be reflected in the REAMDE.md

**Comment by oss6** [comment]
Hi all, I tested the changes with the COVID-19 Data Portal, and everything seems to be fine :+1:

**Comment by khawkins98** [comment]
> Hi all, I tested the changes with the COVID-19 Data Portal, and everything seems to be fine 👍

Nice. I'm hoping to get the readme updated soon.

**Comment by khawkins98** [comment]
I've updated the documentation and made a few small changes to the way event messages are sent so the data format remains consistent in UA.

I believe this should now be ready for merging.

For the amount of effort that went into this PR, the diff reads quite straighforward!

**Review by sandykadam — APPROVED** [review]

**Pull request merged**